File: solutions/python/camicia/2/camicia.py
pay_cards = {'J':1, 'Q':2, 'K':3, 'A':4}

def to_values(hand):
    value = lambda x: pay_cards[x] if x in pay_cards.keys() else 0
    return map(value, hand)    
        
def simulate_game(player_a, player_b):    
    player = [player_a, player_b]
    pile = []
    cards_played = 0
    tricks = 0
    turn = 0
    debt = 0
    seen = set()
    status = ""
    
    while True:
        handA = list(to_values(player[0]))
        handB = list(to_values(player[1]))
        
        if not pile:
            state = (tuple(handA), tuple(handB), turn)
            if state in seen:
                status = "loop"
                break
            else:
                seen.add(state)        
        
        if not player[turn]: 
            if pile!=[]:
                tricks +=1                
            status = "finished"
            break            
         
        current = player[turn].pop(0)        
        pile.append(current)  
        cards_played += 1

        if current in pay_cards.keys():
                debt = pay_cards[current]      
        else:        
             if debt > 0:
                debt -= 1
                if not debt:
                    player[1-turn].extend(pile)
                    pile = []
                    tricks += 1
                    debt = 0  
                    
                    if not player[0] or not player[1]:
                        status = "finished"
                        break
                        
                # if still paying, keep playing 
                else:
                    continue 
        turn = 1 - turn
        
    return {
        "status": status,
        "cards": cards_played,
        "tricks": tricks
    }

File: solutions/dart/camicia/1/lib/camicia.dart
class Camicia {

 int valOfCard(String card){
    switch(card){
      case 'J' : return 1;
      case 'Q' : return 2;
      case 'K' : return 3;
      case 'A' : return 4;
      default : return 0;
    }
  }

  // decks with non-paying cards replaced by '0' 
  List<String> deckOnlyPay(List<String> deck){
    return deck.map((c)=>valOfCard(c)>0 ? c : '0').toList();
  }
    
  Map<String, dynamic> simulateGame(List<String> playerA, List<String> playerB){
        List<String> pile = <String>[];    
        Set<String> seen = <String>{};
        int tricks = 0; 
        int cardsPlayed = 0;
        int debt=0;
        String status = "";    
        List<List<String>> players = [playerA, playerB];
        while(true){     
          var player = players[0];
          var other = players[1];

          if(player.isEmpty){
            tricks += (pile.isEmpty ? 0 : 1);
            status = "finished";
            break;
          }

        // check for looping 
          if(pile.isEmpty){ 
            var state = '${deckOnlyPay(playerA)}|${deckOnlyPay(playerB)}|$debt';
            if (seen.contains(state)){
              status = "loop";
              break;
            } else seen.add(state);
          }         

          // play a card
          final current = player.removeAt(0);
          pile.add(current);
          cardsPlayed += 1;

          if(valOfCard(current)>0){
            // penality card, (re)set debt
            debt = valOfCard(current);
          }
          else {
            if(debt>0){ //there's a debt, pay
              debt -= 1;
              if(debt==0){
                other.addAll(pile);
                pile.clear();
                tricks += 1;
                if(player.isEmpty || other.isEmpty){
                  status = "finished";
                  break;
                } 
              }       
              else continue; // still paying, keep on playing
            }            
          }
          players = [other, player]; //next turn
        }
      return {"status" : status,
              "cards" : cardsPlayed,
              "tricks" : tricks};
          
  }
}
